### PR TITLE
Track a dirty area for in-memory bitmaps

### LINKED ISSUE #1981: Add dirty tracking to bitmap
Changing a bitmap doesn't refresh the screen like it should.

### PR DESCRIPTION
**Pull request by tannewt**
This fixes the bug that bitmap changes do not cause screen updates
and optimizes the refresh when the bitmap is simply shown on the
screen. If the bitmap is used in tiles, then changing it will
cause all TileGrids using it to do a full refresh.

Fixes #1981

### CODE REVIEW

File: shared-module/displayio/Bitmap.c [hunk 1]
```diff
@@ -104,6 +109,26 @@ void common_hal_displayio_bitmap_set_pixel(displayio_bitmap_t *self, int16_t x,
     if (self->read_only) {
         mp_raise_RuntimeError(translate("Read-only object"));
     }
+    // Update the dirty area.
+    if (self->dirty_area.x1 == self->dirty_area.x2) {
+        self->dirty_area.x1 = x;
+        self->dirty_area.x2 = x + 1;
```
Comment (dhalbert) [inline shared-module/displayio/Bitmap.c]: Is it ok if .x2 or .y2 are out of bounds with respect to the bitmap? You don't do a `min(x +1, size_x)` kind of range limiting.

File: shared-module/displayio/TileGrid.c [hunk 2]
```diff
@@ -458,8 +471,21 @@ displayio_area_t* displayio_tilegrid_get_refresh_areas(displayio_tilegrid_t *sel
         return &self->current_area;
     }
 
-    // We must recheck if our sources require a refresh because needs_refresh may or may not have
-    // been called.
+    // If we have an in-memory bitmap, then check it for modifications.
+    if (MP_OBJ_IS_TYPE(self->bitmap, &displayio_bitmap_type)) {
+        displayio_area_t* refresh_area = displayio_bitmap_get_refresh_areas(self->bitmap, tail);
+        if (refresh_area != tail) {
+            // Special case a TileGrid that shows a full bitmap and use it's
```
Comment (dhalbert) [inline shared-module/displayio/TileGrid.c]: its

**Comment by dhalbert** [comment]
Build failed, just looks like `make translate` needed.

File: shared-module/displayio/Bitmap.c (hunk 1 above)
Comment (tannewt) [reply shared-module/displayio/Bitmap.c]: The range limiting is done in shared-bindings so it'll be fine here.

**Comment by tannewt** [comment]
Ok, @dhalbert. It should be ready for re-review and pass Travis 🤞 .

**Comment by dhalbert** [comment]
Is this going to be 4.1.0-rc.1, since it was merged into 4.1.x?

**Pull request merged**